File: LLM-Translations/GPT-4/avatar/Python/C/atcoder_ABC150_D.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>

int gcd(int a, int b) {
    return b == 0 ? a : gcd(b, a % b);
}

int main() {
    int n, num;
    scanf("%d %d", &n, &num);
    int num_set[n];
    for (int i = 0; i < n; i++) {
        scanf("%d", &num_set[i]);
    }
    
    int two_times_set[30];
    for (int i = 0; i < 30; i++) {
        two_times_set[i] = 0;
    }
    
    bool more_than_one = false;
    
    for (int i = 0; i < n; i++) {
        for (int j = 1; j < 30; j++) {
            num_set[i] /= 2;
            if (num_set[i] % 2 != 0) {
                two_times_set[j] = 1;
                break;
            }
        }
        
        int count = 0;
        for (int k = 0; k < 30; k++) {
            if (two_times_set[k] == 1) {
                count++;
            }
        }
        
        if (count != 1) {
            printf("0\n");
            more_than_one = true;
            break;
        }
    }
    
    if (!more_than_one) {
        int lcm = num_set[0];
        for (int i = 1; i < n; i++) {
            lcm = lcm * num_set[i] / gcd(lcm, num_set[i]);
        }
        printf("%d\n", (num - lcm / 2) / lcm + 1);
    }
    
    return 0;
}
// End of Code

File: LLM-Translations/GPT-4/avatar/Python/Java/atcoder_ABC150_D.java
import java.util.*;
import java.math.*;

public class atcoder_ABC150_D {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int num = sc.nextInt();
        sc.nextLine();
        String[] input = sc.nextLine().split(" ");

        HashSet<Integer> num_set = new HashSet<>();
        for (String s : input) {
            num_set.add(Integer.parseInt(s));
        }

        HashSet<Integer> two_times_set = new HashSet<>();
        for (int i : num_set) {
            for (int j = 1; j < 30; j++) {
                i /= 2;
                if (i % 2 != 0) {
                    two_times_set.add(j);
                    break;
                }
            }
            if (two_times_set.size() != 1) {
                System.out.println(0);
                break;
            }
        }

        if (two_times_set.size() == 1) {
            ArrayList<Integer> num_list = new ArrayList<>(num_set);
            int lcm = num_list.get(0);
            for (int i = 1; i < num_list.size(); i++) {
                lcm = lcm * num_list.get(i) / BigInteger.valueOf(lcm).gcd(BigInteger.valueOf(num_list.get(i))).intValue();
            }
            System.out.println((num - lcm / 2) / lcm + 1);
        }
    }
}
// End of Code
